elytron/ELY-2078-encryption-for-FilesystemSecurityRealm.adoc

```diff
@@ -0,0 +1,214 @@
+        <attribute name="RUxZAUMQl1xAhDoJdsMnk7ggw66zyClhUzzB79p859BkQrXazzo=" value="RUxZAUMQCV32L5vunNfLMCISdjyn7myJ0saJdAWreKq1BjnGjfU="></attribute>
+        <attribute name="RUxZAUMQ7t3uOr63lNMwD+Ljsq68ONPwGYdtbVLtdmbY6stZIjg=" value="RUxZAUMQ41IKUmpazIiXd4wQa67I+y35TLtssweZ/1aci7w71ag="></attribute>
+        <attribute name="RUxZAUMQDIjiDvRMr9sOCHnCyh8/uwytBJZOfgCsH0k0itNwcNA=" value="RUxZAUMQi3C21O9+je0b9kXHIsQyoiuSYYafYAFKS3h7iPM9+OE="></attribute>
+    </attributes></identity>
+```
+=== Integrity of Filesystem Realm
+The two approaches being considered are between signing each identity file, and just calculating the checksum of each identity file.
+
+The advantage of using just a straightforward checksum using a hash algorithm is that it will not be resource heavy, as an algorithm such as SHA256 or MD5 is very efficient, although it comes with a caveat that it is possible to produce collisions (false positivies) due to the lack of complexity of the algorithm. The disadvantage of using a simple checksum is that if an authorized user were to gain access to the master list of checksums that is used to verify each identity when accessed, they could change the checksum rendering that identity to be useless, until that checksum is amended. Is it also possible for the unauthorized user to modify the identity file, calculate the new checksum, and update the checksum list with an updated one. This would be very detrimental as we have no way of knowing that the file has been modified and is no longer valid.
+
+If the approach of signing a file was chosen, it would require an asymmetric key pair, a public key and secret key. This would involve another credential store and a method to ensure only the correct party has access to the secret key.
+
+It could be made that the public key is available for everyone to use. This means anyone can decrypt the signature to get the hash of the identity, but only people with access to the secret key for the encryption of the realm would have access to the private key that encrypts the hash of each identity. This way an unauthorized user could view the hash via the public key, but cannot change the hash/signature of the identity.
+
+Since signing each xml identity file is more secure, this was the approach that will be taken.
+The details about adding a signature to an XML file is described here and also has the XSD schema: https://www.w3.org/TR/xmldsig-core2/#sec-xsdSchema
+
+==== Hard Requirements
+In the FileSystemSecurityRealm in Elytron, 2 new optional parameters were introduced to the constructor that accepts a PublicKey, and PrivateKey. Passing in both a PublicKey and PrivateKey indicates that the user wants to enable filesystem integrity to the realm. If only one of the two keys is set, it is ignored completely (although this should never occur due to the implementation done in wildfly-core).
```

Comment:
If only one of the two keys is set, the builder's build method could throw an IllegalArgumentException.

Reply:
Should I do any validation in wildfly core to ensure that both keys are passed in, or leave it to the elytron filesystem realm builder to throw the error?

Reply:
It would be good to add validation in WildFly Core as well.

Reply:
Sounds good, I updated that here and will make that change to Elytron and Core right now